cilium/CFP-28427-fqdn-selectors.md

```diff
@@ -0,0 +1,277 @@
+ipc-- UpdateIdentities() -->sc
+nm-- UpdateFQDNSelector() -->sc
+pol-- SyncPolicyMap() -->bpf
+```
+
+#### The SelectorCache
+
+The SelectorCache maintains a list of all "active" selectors for policies. If, for example, there is a network policy that grants access to label `team: database`, this selector will "live" in the SelectorCache. The policy will ask the SelectorCache for all identities that match this selector. Furthermore, the policy will "register" for incremental updates to this selector.
+
+#### NameManager
+
+The NameManager maintains two key lists:
+1. The mapping from DNS names to IPs
+2. The set of active FQDN selectors
+
+As DNS responses come in, the NameManager determines if the name is selected by any ToFQDN policies. If so, it inserts the IP in to the ipcache (if needed) and updates the relevant ToFQDN selectors.
+
+#### IPCache
+
+The IPCache stores the mapping from prefix to labels. It also allocates identities for labels, and manages the mapping from prefix -> numeric identity.
+
+### Incremental updates
+
+1. An identity is allocated for the IP:
+  a. `ipc.UpsertPrefixes()` is called, associating the complete set of CIDR labels with this IP address.
+  b. The IPCache determines the IP's full set of labels, allocates an identity, and informs the SelectorCache of this new IP
+2. The selector is updated:
+  a. The NameManager calls `sc.UpdateFQDNSelector()`, updating the set of selected IPs for this given FQDNSelector
+  b. The SelectorCache determines the complete set of identities selected
+  c. Any differences are pushed to network policies
+3. Endpoints are updated
+  Any pending incremental updates are batched up and applied to the endpoints' BPF PolicyMap. If needed, Envoy is also updated
+ 
+Only at this point can the DNS packet be returned to the waiting endpoint.
+
+## Performance bottlenecks
+
+### Many identities
+
+Some DNS names (e.g S3) will return many different IP addresses. S3 even has a TTL of 5 seconds, so the set of IPs will quickly be very broad. This causes a large number of identities to be allocated -- one per IP. These identities then need to plumbed through to all endpoints.
+
+### Label sizes
+
+CIDR labels, especially IPv6 labels, are large. For example, the full set of labels for the CIDR `2001:db8:cafe::cab:4:b0b:0/112` is 2KiB. Every extra-cluster identity always has the full set of CIDR identities calculated for it.
+
+----
+
+## Proposal
+
+This CFP proposes 3 key changes. Together, they preserve correctness while radically restructuring how prefixes are selected.
+
+1: Labels on prefixes propagate downwards. That is to say, if prefix `192.0.0.0/22` has label `foo:bar`, then prefix `192.0.2.0/24` also has label `foo:bar`, unless overridden. The IPCache aggregates labels accordingly.
+2: Only selected CIDR labels are inserted in to the ipcache. If a selector references CIDR `192.0.2.0/24`, only the prefix -> label mapping `192.0.2.0/24 -> cidr:192.0.2.0/24` is added to the IPCache.
```

Comment:
sidenote: To implement CIDR policy, the Identity must be allocated after the CIDR policy is processed and that Identity must be (a) inserted into the bpf ipcache and (b) the Identity must be evaluated against the policies applied to local endpoints and inserted into the corresponding relevant policymaps. This process is unrelated to the FQDNs functionality described in this CFP, but already occurs and must still occur to correctly implement the CIDR policy feature.

Reply:
Does ordering matter? I don't think so; I'm pretty sure the SelectorCache does the right thing. What if a CIDR policy selects a `/32` that has already been allocated by FQDN policy? It should work fine.

Reply:
There's a few permutations to consider (sample, not comprehensive):

- No policy exists, create a new CIDR policy for this IP
- Presuppose a broader CIDR policy, then a new more specific CIDR policy is created
- Presuppose a more specific CIDR policy, then a new broader CIDR policy is created
- Presuppose an FQDN policy, then a CIDR policy with the same IP is created

I think that each of these cases more or less boils down to the same set of operations that need to occur in the correct order to prevent connection disruption due to (temporary) packet loss:

1. Cilium processes the CIDR policy
2. Cilium plumbs the policy down into the PolicyRepository, triggers Endpoint regeneration, leading to generation of the SelectorPolicy and generally preparing the SelectorCache so that when there *is* an identity, we can figure out how to plumb the policymaps correctly.
3. Somewhere along the way Cilium needs to inform the IPCache about the IP<->label mapping
4. The IPCache will react to the label mapping and allocate an Identity for the IP.
   1. If this happens before (2), then this will plumb the ipcache bpf map without updating the policymaps, and may lead to packet loss. Newly arriving traffic will hit the datapath, hit the ipcache lookup, resolve to the new identity, then subsequently hit the policy lookup and the policymap is not prepared for this new Identity which leads to drops.
      1. Given that this happens first, once (2) eventually _does_ occur, it will automatically calculate the new policy against the latest set of Identities including the one just allocated here in step (4) and the policymaps will converge to be "eventually consistent" and allow the traffic again. However by that point we've already disrupted the connection.
   2. If (2) happens before (4), then the ipcache will ensure correct ordering by first notifying the selectorcache about the newly allocated Identity, and waiting for the corresponding policymaps to be plumbed, then subsequently update the bpf ipcache.

Looking through these steps, if there is another policy already in place such as a broader CIDR policy or an FQDN policy, then step (2) is already in place and so we can guarantee case (2.ii) and all is fine. The tricky cases are when there is no specific policy for this IP yet, when the newly created policy is broader than the existing one, or for extra spice when there are overlapping properties (consider for instance a policy that is broader on L3 but more specific on L4 with L7 aspects or auth). Let's say for example if there is a /32 policy already, then we create a new /24 policy on port 80 with L7 filtering. Given the dynamic Identity creation proposed in this PR, there will already be Identity X for the /32 (no /24 label), then once all is processed there must be two identities, a new Identity X′ (/24 and /32 label) and Y (/24 label), the SelectorPolicy must be recalculated to account for the distinct subsets of traffic {/32 and port 80, /32 and not port 80, not /32 and port 80}, and of course the new bpf ipcache and policymap content must be calculated based on the new SelectorPolicy and Identities.

I think we actually _don't_ provide this level of ordering guarantee when calculating CIDR policies today; we only guarantee that policy calculation is _triggered_ before informing the ipcache, but don't guarantee that it _completes_ before informing the ipcache:

https://github.com/cilium/cilium/blob/12e3ae9936bd82924a995cbf2a2eb6284d6db0cb/daemon/cmd/policy.go#L464-L471

Reply:
For existing CIDR selectors, if a given IP changes identities due to a label being added, that change can be rolled out without dropping traffic. This is because the ipcache is very careful about the order in which it interlaves operations.

For *new* selectors, traffic may be dropped. This is unrelated to the proposed FQDN refactor. I actually have a branch that fixes this, but it's not yet ready for merging.